### PR TITLE
Add eval fine tuning

### PR DESCRIPTION
**Pull request by gkumbhat**
Closes: #75 

### Changes
- Update transformers to latest version that enables bugfix for Trainer prediction
- Add [samsum](https://huggingface.co/datasets/samsum) and [billsum](https://huggingface.co/datasets/billsum) dataset from huggingface to test out summarization
- Change default input text length to support summarization
- Fix prediction generation batch decoding to single output mapping issue
- Add evaluation in fine-tuning notebook

### Test
1. Unit test: `tox -e py39 -- tests`
2. Evaluation example:
    ```python
   ALLOW_DOWNLOADS=true python3 run_fine_tuning.py --dataset "billsum"  --model_name t5-base --num_epochs 4 --output_dir tmp/with_loss_log --batch_size=8 --accumulate_steps 16 --max_source_length 1024 --metric rouge --torch_dtype float16
    ```

### CODE REVIEW

File: examples/utils.py [hunk 1]
```diff
@@ -80,9 +80,10 @@ def get_distributed_model(model_path):
         kill_tgis_container_if_exists()
 
     # TODO: Enforce validation that TGIS is mounting the same model type
-    _CONFIGURED_BACKENDS.clear()
     caikit.configure(
-        config_dict={"module_backends": {"priority": [TGISBackend.backend_type]}}
+        config_dict={
+            "module_backends": {"load_priority": [{"type": TGISBackend.backend_type}]}
```
Comment (evaline-ju) [inline examples/utils.py:85]: I wanted to try to run this but noticed these config changes and the `caikit.core.module_backends.module_backend_config` import appear to work with a `caikit` version prior to `0.10.0` (the current lower bound) - what version was this tested with? The caikit module backends appeared reorganized for `0.10.0` - https://github.com/caikit/caikit/pull/275

File: examples/utils.py [hunk 2]
```diff
@@ -226,6 +227,46 @@ def to_generation_fmt(x):
...
+    dataset = datasets.load_dataset("billsum", split="ca_test")
+    train_test_dataset = dataset.train_test_split(test_size=0.2)
+    # # Split the 10% test + valid into half test, half valid
+    # test_valid = train_test_valid_dataset['test'].train_test_split(test=0.5)
+
+    build_stream = lambda split: caikit.core.data_model.DataStream.from_iterable(
+        [to_generation_fmt(datum) for datum in train_test_dataset[split]]
+    )
+    train_stream = build_stream("train")
+    validation_stream = build_stream("test")
```
Comment (evaline-ju) [inline examples/utils.py:245]: wanted to confirm if the "test" here is intentional vs. "validation"? same for the above financial phrasebook and not accidental copypaste

File: caikit_nlp/modules/text_generation/fine_tuning.py [hunk 3]
```diff
@@ -119,6 +120,23 @@ def train(
             shuffle=True,
         )
 
+        ### Dtype based processing
+        # NOTE: Following is not exhaustive list of all parameters
+        # for all dtypes
+        if torch_dtype == torch.float16:
```
Comment (evaline-ju) [inline caikit_nlp/modules/text_generation/fine_tuning.py:126]: would it be possible to add small unit test(s) for these code paths or is the overhead for train/run pretty high right now even for unit tests?

Comment (gkumbhat) [reply caikit_nlp/modules/text_generation/fine_tuning.py:126]: actually anything half precision is only available on GPUs, so the current unit tests setup won't work for this

File: examples/utils.py (hunk 1 above)
Comment (gkumbhat) [reply examples/utils.py:85]: this was tested on `0.9.1`. Need to get this updated.

File: examples/utils.py (hunk 2 above)
Comment (gkumbhat) [reply examples/utils.py:245]: Yep, since this dataset doesn't have "validation" set

**Review by evaline-ju — APPROVED** [review]
LGTM - anticipating the examples scripts will be fixed in https://github.com/caikit/caikit-nlp/pull/90

**Review by evaline-ju — APPROVED** [review]

**Pull request merged**